game.cpp

```diff
@@ -1808,6 +1844,83 @@ void test_game() //!OCLINT tests may be many
   }
   #endif // FIX_ISSUE_478
 
+  #define FIX_ISSUE_524
+  #ifdef FIX_ISSUE_524
+  {
+    // do_action will change a player's action flag
+    {
+      game g;
+      g.do_action(1, action_type::accelerate_backward);
+      assert(g.get_player(1).get_action() == action_type::accelerate_backward);
+      g.do_action(1, action_type::accelerate_forward);
+      assert(g.get_player(1).get_action() == action_type::accelerate_forward);
+      g.do_action(1, action_type::none);
+      assert(g.get_player(1).get_action() == action_type::none);
+      g.do_action(1, action_type::shoot);
+      assert(g.get_player(1).get_action() == action_type::shoot);
+      g.do_action(1, action_type::shoot_stun_rocket);
+      assert(g.get_player(1).get_action() == action_type::shoot_stun_rocket);
+      g.do_action(1, action_type::turn_left);
+      assert(g.get_player(1).get_action() == action_type::turn_left);
+      g.do_action(1, action_type::turn_right);
+      assert(g.get_player(1).get_action() == action_type::turn_right);
+    }
+
+    // reset_player_action() can reset all players' action flags
+    {
+        game g;
+        for (auto i = 0; i < static_cast<int>(g.get_v_player().size()); ++i)
+        {
+            g.do_action(i, action_type::accelerate_forward);
+            assert(g.get_player(i).get_action() == action_type::accelerate_forward);
+        }
+        g.reset_player_action();
+        for (auto i = 0; i < static_cast<int>(g.get_v_player().size()); ++i)
+        {
+            assert(g.get_player(i).get_action() == action_type::none);
+        }
+    }
+
+    // apply_inertia() works properly
```

Comment:
Ah, there is the inertia. This is a looooong test, it used to be shorter. I would enjoy it to be short again. If you test two different things, just use two different tests :-)

Reply:
The test is long because I wish to test two things that one happens after another. 1) When there is key input of move forward or backward, `apply_inertia()` does nothing; 2) When there is no such input at the next tick, `apply_inertia()` decelerates players. I removed some safety nets (aka redundant tests) to make it shorter : )